### PR TITLE
Clone empty blocks before def-use analysis

### LINKED ISSUE #3650: Defuse analysis fails with "Overwriting definitions"
The attached file: [defuse_fail.p4.txt](https://github.com/p4lang/p4c/files/9914186/defuse_fail.p4.txt) triggers a failed BUG_CHECK during def-use analysis:

`Compiler Bug: Overwriting definitions at <BlockStatement>(1396)`

I checked the IR before def-use is invoked - the problematic BlockStatement is reachable on two control-flow program paths, but def-use seems to assume the IR doesn't share nodes. The node sharing is created by the inlining pass (running before defuse since https://github.com/p4lang/p4c/pull/3591). Inlining substitutes both C1.apply method calls with C1 body, but the body isn't deep-copied, hence the two places where C1 is inlined share part of the IR subtree.

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Signed-off-by: Mihai Budiu <mbudiu@vmware.com>
Fixes #3650 
This was harder to fix than I expected. The visitor infra is too smart for its own good: it wouldn't duplicate empty blocks since it realizes they are identical. So you have to add annotations to empty blocks to make then different, then remove them again.

This also uncovered a bug in the frontend where RemoveUnusedDeclarations was called instead of RemoveAllUnusedDeclarations. That won't happen again, since I made the constructor private.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
@fruffy this fails in testgen with the following error:
`Required regular expression not found. Regex=[Duplicate objects of type '.*' with name '.*'`

**Comment by fruffy** [comment]
It looks like this also fixes #3620.

**Comment by mbudiu-vmw** [comment]
If it fixes #3620 it is a coincidence. That one I still have to work on.

**Comment by fruffy** [comment]
> If it fixes #3620 it is a coincidence. That one I still have to work on.

We can leave the issue open then, but I do not have a reproducing test for this right now. Either way, it must be removed from the xfails here.

File: frontends/p4/simplifyDefUse.h [hunk 1]
```diff
@@ -48,16 +48,49 @@ class DoSimplifyDefUse : public Transform {
...
 class SimplifyDefUse : public PassManager {
     class Cloner : public ClonePathExpressions {
      public:
         Cloner() { setName("Cloner"); }
         const IR::Node* postorder(IR::EmptyStatement* stat) override {
             // You cannot clone an empty statement, since
             // the visitor claims it's equal to the original one.
-            // So we cheat and make an empty block.
+            // So we convert it to an empty block.
             return new IR::BlockStatement(stat->srcInfo);
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply frontends/p4/simplifyDefUse.h:75]: I do produce an empty block, but no annotation is needed here.
There is really no trick, except I need each statement to have its own IR node.

File: frontends/common/programMap.h [hunk 2]
```diff
@@ -56,6 +57,11 @@ class ProgramMap : public IHasDbPrint {
         program = node->to<IR::P4Program>();
         LOG2(mapKind << " updated to " << dbp(node));
     }
+    void clear() {
+        // This ensures that a clear map is never up-to-date,
+        // since the 'fake' node cannot appear in a program.
+        program = fake;
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply frontends/common/programMap.h:63]: I would not call it a hack.
Each map stores the program that it was generated on. If the program hasn't changed, the map is good.
But an empty map is never up-to-date.

**Comment by mbudiu-vmw** [comment]
> > If it fixes #3620 it is a coincidence. That one I still have to work on.
> 
> We can leave the issue open then, but I do not have a reproducing test for this right now. Either way, it must be removed from the xfails here.

It should be easy to create a reproduction by having a metadata and a header structure with a common field name.

**Review by vhavel — APPROVED** [review]
LGTM

**Pull request merged**